packages/opentelemetry-sdk-trace-web/src/utils.ts

```diff
@@ -125,6 +129,139 @@ export function addSpanNetworkEvents(
+ *
+ * The fetch API is kinda messy: there are a couple of ways the body can be passed in.
+ *
+ * In all cases, the body param can be some variation of ReadableStream,
+ * and ReadableStreams can only be read once! We want to avoid consuming the body here,
+ * because that would mean that the body never gets sent with the actual fetch request.
+ *
+ * Either the first arg is a Request object, which can be cloned
+ *   so we can clone that object and read the body of the clone
+ *   without disturbing the original argument
+ *   However, reading the body here can only be done async; the body() method returns a promise
+ *   this means this entire function has to return a promise
+ *
+ * OR the first arg is a url/string
+ *   in which case the second arg has type RequestInit
+ *   RequestInit is NOT cloneable, but RequestInit.body is writable
+ *   so we can chain it into ReadableStream.pipeThrough()
+ *
+ *   ReadableStream.pipeThrough() lets us process a stream and returns a new stream
+ *   So we can measure the body length as it passes through the pie, but need to attach
+ *   the new stream to the original request
+ *   so that the browser still has access to the body.
+ *
+ * @param body
+ * @returns promise that resolves to the content length of the body
+ */
+export function getFetchBodyLength(...args: Parameters<typeof fetch>) {
+  if (args[0] instanceof URL || typeof args[0] === 'string') {
+    const requestInit = args[1];
+    if (!requestInit?.body) {
+      return Promise.resolve();
+    }
+    if (requestInit.body instanceof ReadableStream) {
+      const { body, length } = _getBodyNonDestructively(requestInit.body);
+      requestInit.body = body;
+
+      return length;
+    } else {
+      return Promise.resolve(getXHRBodyLength(requestInit.body));
+    }
+  } else {
+    const info = args[0];
+    if (!info?.body) {
+      return Promise.resolve();
+    }
+
+    return info
+      .clone()
+      .text()
+      .then(t => t.length);
+  }
+}
+
+/**
+ * Helper function to determine payload content length for XHR requests
+ * @param body
+ * @returns content length
+ */
+export function getXHRBodyLength(
+  body: Document | XMLHttpRequestBodyInit
+): number {
+  if (typeof Document !== 'undefined' && body instanceof Document) {
+    return new XMLSerializer().serializeToString(document).length;
+  }
+  // XMLHttpRequestBodyInit expands to the following:
+  if (body instanceof Blob) {
+    return body.size;
+  }
+
+  // ArrayBuffer | ArrayBufferView
+  if ((body as any).byteLength !== undefined) {
+    return (body as any).byteLength as number;
+  }
+
+  if (body instanceof FormData) {
+    // typescript doesn't like it when we pass FormData into URLSearchParams
+    // even though this is actually totally valid
+    return new URLSearchParams(body as any).toString().length;
+  }
+
+  if (body instanceof URLSearchParams) {
+    return body.toString().length;
+  }
+
+  if (typeof body === 'string') {
+    return body.length;
```

Comment:
Javascript uses [UTF-16](https://en.wikipedia.org/wiki/UTF-16) encoding for strings, so what's returned by `body.length` isn't actually the number of bytes for a given string, it's the number of UTF-16 code units used to represent the string.

Helpful links:
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#utf-16_characters_unicode_code_points_and_grapheme_clusters
* https://github.com/smithy-lang/smithy-typescript/blob/main/packages/util-body-length-browser/src/calculateBodyLength.ts#L12-L21

Reply:
Thanks, this is a good point! 

The sample you linked to from Smithy checks if TextEncoder is available and falls back to some manual math if it isn't. I opted to just go straight to TextEncoder because [caniuse says it's available](https://caniuse.com/?search=textencoder) on all browser platforms that we support.

Reply:
I had the same initial thought when I saw that as well, but my thinking was that while it would definitely more maintainable (and clear) to use `TextEncoder` that it'd probably be wasteful to encode the string to UTF-8 just to count its length. Depending on the request body size, it might be a meaningful amount of additional memory.

Reply:
I had the same thought when I saw that as well, but my reasoning was that while it would definitely more maintainable (and clear) to use `TextEncoder`, that it'd probably be wasteful to encode the string + allocate a separate Uint8Array just to count its length. Depending on the request body size, it might be a meaningful amount of additional memory.

Reply:
That's fair but I don't expect this case to hit very many large payloads. Most large payloads (ie. file uploads) will get handled separately, either as a ReadableStream (where we're using pipeThrough) or as FormData (where we'll get the byte length directly).

Reply:
@MustafaHaddara For the most part I think that would be absolutely true! But I also wouldn't underestimate the potential for people to send large JSON payloads (either intentionally or unintentionally). 

As a contemporary example, OpenAI's chat API accepts user image input as ["[e]ither a URL of the image or the base64 encoded image data."](https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages), so you can image that for a large image this might amount to some meaningful overhead if they send images inline.

Given some of the other inherent issues for calculating content length size with `FormData`, would you be open to allowing users to specify their own optional `getXHRBodyLength` (or maybe exposed as `calculateBodyLength`) function?

For context, I'm hoping to use this functionality in [Browser Extension for OpenTelemetry](https://github.com/tbrockman/browser-extension-for-opentelemetry), and I'd like to limit the overhead as much as possible when it already involves injecting big blobs of Javascript into pages which might be making requests that users might not have much control over. This way, I can supply my own implementation (which may be less maintainable/correct/costly) and also experiment with browser-specific calculations, without any maintenance burden being placed on the project.